### PR TITLE
feat: [CDS-83774] Update YAML schema (v0 and v1) for K8s Canary step to add traffic routing

### PR DESCRIPTION
**Pull request by alebuh**
Tech spec used for defining schemas https://harness.atlassian.net/wiki/spaces/CDP/pages/21455503361/Tech+Spec+Percentage+Based+Traffic+Shifting+Steps+for+Canary+and+Blue+Green+for+Kubernetes+Deployment+Types#SMI-provider-supported-rules
Changes for v0 and v1 are the same. Once we start the migration of k8s steps, those schemas will be updated

### CODE REVIEW

File: v0/pipeline/steps/cd/k8s-istio-provider-spec.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+title: IstioProviderSpec
+type: object
+required:
+  - gateways
```
Comment (markoharness) [inline v0/pipeline/steps/cd/k8s-istio-provider-spec.yaml:4]: we should make gateways and hosts fields not mandatory

Comment (alebuh) [reply v0/pipeline/steps/cd/k8s-istio-provider-spec.yaml:4]: should we update tech spec in that case? 
Screenshot 2023-11-10 at 12 21 19

File: v0/pipeline/steps/cd/k8s-istio-provider-spec.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+properties:
+  gateways:
+    type: array
+    items:
+      type: string
+  hosts:
+    type: array
+    items:
+      type: string
+  routes:
```
Comment (markoharness) [inline v0/pipeline/steps/cd/k8s-istio-provider-spec.yaml:17]: we should add routes and destinations to a sort of base object and here call "allOf" 

something like


/schema/base
{
  "type": "object",
  "properties": {
    "routes": { ... },
    "destinations": { ... }
  }
}


{
  "allOf": [{ "$ref": "/schema/base" }],
  "properties": {
    "gateways": { ... }
  }
}

File: v0/pipeline/steps/cd/k8s-smi-provider-spec.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+title: SMIProvider
+type: object
+required:
+  - routes
+  - destinations
+properties:
+  rootService:
+    type: string
+  routes:
```
Comment (markoharness) [inline v0/pipeline/steps/cd/k8s-smi-provider-spec.yaml:9]: same here

File: v0/pipeline/steps/cd/k8s-traffic-routing-destination.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+title: TrafficRoutingDestination
+type: object
+required:
+  - destination
+properties:
+  destination:
+    type: object
+    required:
+      - host
+      - weight
```
Comment (markoharness) [inline v0/pipeline/steps/cd/k8s-traffic-routing-destination.yaml:10]: weight should not be required

File: v1/pipeline/steps/cd/k8s-istio-provider-spec.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+title: IstioProviderSpec
+type: object
+required:
+  - gateways
+  - hosts
```
Comment (markoharness) [inline v1/pipeline/steps/cd/k8s-istio-provider-spec.yaml:5]: same here

File: v1/pipeline/steps/cd/k8s-istio-provider-spec.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+properties:
+  gateways:
+    type: array
+    items:
+      type: string
+  hosts:
+    type: array
+    items:
+      type: string
+  routes:
```
Comment (markoharness) [inline v1/pipeline/steps/cd/k8s-istio-provider-spec.yaml:17]: same here

File: v1/pipeline/steps/cd/k8s-smi-provider-spec.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+title: SMIProvider
+type: object
+required:
+  - routes
+  - destinations
+properties:
+  rootService:
+    type: string
+  routes:
```
Comment (markoharness) [inline v1/pipeline/steps/cd/k8s-smi-provider-spec.yaml:9]: same here

File: v1/pipeline/steps/cd/k8s-traffic-routing-destination.yaml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+title: TrafficRoutingDestination
+type: object
+required:
+  - destination
+properties:
+  destination:
+    type: object
+    required:
+      - host
+      - weight
```
Comment (markoharness) [inline v1/pipeline/steps/cd/k8s-traffic-routing-destination.yaml:10]: same here

File: v0/pipeline/steps/cd/k8s-istio-provider-spec.yaml (hunk 1 above)
Comment (alebuh) [reply v0/pipeline/steps/cd/k8s-istio-provider-spec.yaml]: done

File: v1/pipeline/steps/cd/k8s-istio-provider-spec.yaml (hunk 6 above)
Comment (alebuh) [reply v1/pipeline/steps/cd/k8s-istio-provider-spec.yaml]: done

File: v1/pipeline/steps/cd/k8s-smi-provider-spec.yaml (hunk 7 above)
Comment (alebuh) [reply v1/pipeline/steps/cd/k8s-smi-provider-spec.yaml]: done

File: v1/pipeline/steps/cd/k8s-traffic-routing-destination.yaml (hunk 8 above)
Comment (alebuh) [reply v1/pipeline/steps/cd/k8s-traffic-routing-destination.yaml]: done

File: v0/pipeline/steps/cd/k8s-smi-provider-spec.yaml (hunk 3 above)
Comment (alebuh) [reply v0/pipeline/steps/cd/k8s-smi-provider-spec.yaml]: done

File: v1/pipeline/steps/cd/k8s-istio-provider-spec.yaml (hunk 5 above)
Comment (alebuh) [reply v1/pipeline/steps/cd/k8s-istio-provider-spec.yaml]: done

File: v0/pipeline/steps/cd/k8s-route-spec.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    type: object
+    required:
+      - type
+      - rules
+    properties:
+      type:
+        type: string
+        enum:
+          - http
+          - tcp
```
Comment (abosii-harness) [inline v0/pipeline/steps/cd/k8s-route-spec.yaml:16]: are we planning to support all these types in first phase? @markoharness

File: v1/pipeline/steps/cd/k8s-traffic-routing-spec.yaml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+required:
+  - spec
+  - provider
+properties:
+  spec:
+    type: object
+  provider:
+    type: string
+    enum:
+      - SMI
```
Comment (abosii-harness) [inline v1/pipeline/steps/cd/k8s-traffic-routing-spec.yaml:12]: any other enums are lowercase, any reason why SMI and Istio follow different convention?

Comment (alebuh) [reply v1/pipeline/steps/cd/k8s-traffic-routing-spec.yaml:12]: no particular reason, I just followed tech spec, but you are right I don't see a reason to not keep it as lowercase

Comment (alebuh) [reply v1/pipeline/steps/cd/k8s-traffic-routing-spec.yaml]: done

File: v0/pipeline/steps/cd/k8s-route-spec.yaml (hunk 9 above)
Comment (markoharness) [reply v0/pipeline/steps/cd/k8s-route-spec.yaml:16]: @alebuh please remove all route type enums except "http". For now lets go only with HTTP.

File: v1/pipeline/steps/cd/k8s-route-spec.yaml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+  route:
+    type: object
+    required:
+      - type
+      - rules
+    properties:
+      type:
+        type: string
+        enum:
+          - http
```
Comment (markoharness) [inline v1/pipeline/steps/cd/k8s-route-spec.yaml:15]: same here

File: v0/pipeline/steps/cd/k8s-traffic-routing-rule.yaml [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+required:
+  - type
+  - spec
+properties:
+  type:
+    type: string
+    enum:
+      - uri
+      - method
+      - header
```
Comment (markoharness) [inline v0/pipeline/steps/cd/k8s-traffic-routing-rule.yaml:12]: Not all of these enums apply for all route types. Can you check how can we be more specific?

File: v1/pipeline/steps/cd/k8s-traffic-routing-rule.yaml [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+type: object
+required:
+  - type
+  - spec
+properties:
+  type:
+    type: string
+    enum:
+      - uri
+      - method
```
Comment (markoharness) [inline v1/pipeline/steps/cd/k8s-traffic-routing-rule.yaml:11]: same here

File: v0/pipeline/steps/cd/k8s-traffic-routing-rule.yaml (hunk 12 above)
Comment (alebuh) [reply v0/pipeline/steps/cd/k8s-traffic-routing-rule.yaml:12]: I was looking into it and didn't find an easy way of doing it through static schema as there is no way of passing some enum value from parent to child schema. To do that we would need to create a specific schema for each Provider - Route - Rule relation. For example Schema for SmiHttpUri, IstioHttpUri, SmiHttpPort, IstioHttpPort... and that would increase complexity a lot. I found this implementation optimal and the idea is to do additional validation on the backend when the time comes.

File: v1/pipeline/steps/cd/k8s-route-spec.yaml (hunk 11 above)
Comment (alebuh) [reply v1/pipeline/steps/cd/k8s-route-spec.yaml:15]: done

File: v1/pipeline/steps/cd/k8s-traffic-routing-rule.yaml (hunk 13 above)
Comment (alebuh) [reply v1/pipeline/steps/cd/k8s-traffic-routing-rule.yaml:11]: done

**Review by ivan-83 — APPROVED** [review]

**Review by markoharness — APPROVED** [review]

**Review by abosii-harness — APPROVED** [review]

**Review by acasian-harness — APPROVED** [review]

**Pull request merged**